File: Java/Queue/src/Queue.java
import java.sql.Array;
import java.util.Arrays;

public class Queue<T> {

    private T[] queue;
    private int capacity;
    private int length = 0;

    public Queue(int capacity) {
        this.queue = (T[]) new Object[capacity];
        this.capacity = capacity;
    }

    public Queue() {
        this(10);
    }

    public String toString() {
        return Arrays.toString(this.queue);
    }

    public int length(){
        return this.length;
    }

    public boolean isEmpty() {
        return this.length == 0;
    }

    private void increaseCapacity(int newCapacity) {
        if (this.length == this.capacity) {
            final T[] newQueue = (T[]) new Object[this.capacity + newCapacity];
            for (int i = 0; i <= this.length - 1; i++) {
                newQueue[i] = this.queue[i];
            }
            this.queue = newQueue;
        }
    }

    public void showNext() {
        System.out.println(this.queue[0]);
    }

    public void add(T item) {
        if (this.length == this.capacity) {
            this.increaseCapacity(10);
        }

        this.queue[this.length] = item;
        this.length++;
    }

    public T getNext() throws EmptyQueueException{
        if(this.queue[0] == null || this.length == 0){
            throw new EmptyQueueException("Queue is empty");
        }

        T getElement = this.queue[0];
        for (int i = 0; i <= this.length - 1; i++) {
            this.queue[i] = this.queue[i + 1];
        }
        return getElement;
    }

    public void resetQueue(){
        this.length = 0;
        this.queue = (T[]) new Object[this.capacity];
    }


}

class EmptyQueueException extends Exception{

    public EmptyQueueException(String message){
        super(message);
    }

}

File: Kotlin/Queue/src/Queue.kt
import com.sun.org.apache.xpath.internal.operations.Bool
import java.util.Arrays

open class Queue<T>(private var capacity: Int) {

    constructor() : this(10)

    private var queue: Array<Any> = Array<Any>(this.capacity, {})
    private var length: Int = 0

    override fun toString(): String {
        return Arrays.toString(this.queue)
    }

    fun isEmpty(): Boolean {
        return length == 0
    }

    fun length(): Int {
        return this.length
    }

    private fun increaseCapacity(newCapacity: Int = 10) {
        lateinit var newQueue: Array<Any>
        if (this.length == this.capacity) {
        newQueue = Array<Any>(this.capacity + newCapacity, {})
            for (i: Int in 0..this.length - 1) {
                newQueue[i] = this.queue[i]
            }
        }
        this.queue = newQueue
    }

    fun showNext() {
        if(this.isEmpty()){
            println("Queue is empty")
        }

        println("Next value: ${this.queue[0]}")
    }


    fun add(item: T) {
        if (this.length == this.capacity) {
            this.increaseCapacity(1)
        }

        this.queue[length] = item as Any
        length++
    }

    fun getNext(): T {
        if(this.queue[0] is Unit || this.length() == 0){
            throw EmptyQueueException()
        }

        val getElement: T = this.queue[0] as T
        for (i: Int in 0..this.length - 1) {
            this.queue[i] = this.queue[i + 1]
        }
        this.length--
        return getElement
    }

    fun resetQueue(){
        this.length = 0
        this.queue = Array<Any>(this.capacity, {})
    }

    class EmptyQueueException() : Exception("Queue is empty")
}